### PR TITLE
#1974: fix various c++17 deprecations and compile errors

### LINKED ISSUE #1974: address C++17 deprecations
**What Needs to be Done?**

- [ ] Replace `std::is_pod` with `std::is_standard_layout && std::is_trivial`
- [ ] Remove `std::iterator`

### PR DESCRIPTION
**Pull request by nmm0**
Fixes #1974

### CODE REVIEW

**Comment by PhilMiller** [comment]
Lol at those compiler errors

File: src/vt/termination/interval/integral_set.h [hunk 1]
```diff
@@ -378,8 +378,7 @@ struct IntegralSetBase {
   // using reference         = DomainT&;
 
   template <typename Impl>
-  struct IntervalSetIter :
-    std::iterator<std::bidirectional_iterator_tag, DomainT, DomainT>
+  struct IntervalSetIter
```
Comment (PhilMiller) [inline src/vt/termination/interval/integral_set.h:381]: I'm not familiar with this bit of C++ evolution - what previously depended on having that inheritance?

File: src/vt/vrt/collection/balance/lb_common.h [hunk 2]
```diff
@@ -76,6 +76,10 @@ static constexpr ElementIDType const no_element_id = elm::no_element_id;
 struct PhaseOffset {
   PhaseOffset() = delete;
 
+  PhaseOffset(int ph, unsigned int sub)
+    : phases(ph), subphase(sub)
+  {}
+
```
Comment (PhilMiller) [inline src/vt/vrt/collection/balance/lb_common.h:82]: What changed with C++17 that we need an explict constructor rather than just an implicit initializer list construction?

**Comment by nmm0** [comment]
> Lol at those compiler errors

:facepalm: maybe I should have tested with C++14.... well I guess I'm going to have to find another way. Maybe the least gross way will be to just do explicit capture for everything, no default capture

File: src/vt/termination/interval/integral_set.h (hunk 1 above)
Comment (nmm0) [reply src/vt/termination/interval/integral_set.h:381]: Yeah I was curious about that too -- I found this explanation: https://www.fluentcpp.com/2018/05/08/std-iterator-deprecated/

tl;dr basically `std::iterator` was just a helper and would take as template parameters the trait types (e.g. value_type). We never needed it since we specify the types directly.

**Comment by PhilMiller** [comment]
> > Lol at those compiler errors
> 
> 🤦 maybe I should have tested with C++14.... well I guess I'm going to have to find another way. Maybe the least gross way will be to just do explicit capture for everything, no default capture

Or wait a bit, and we can fold in "require C++17" with the changes to adapt to it. EMPIRE will be able to accept that in the next few weeks, BVH is already there as I understand it, and I can check on our other app.

File: src/vt/termination/interval/integral_set.h (hunk 1 above)
Comment (PhilMiller) [reply src/vt/termination/interval/integral_set.h:381]: Thanks, I see now.

It seems that the types as specified maybe aren't right, though? They even disagree with the comment just above.

@lifflander ?

Comment (nmm0) [reply src/vt/termination/interval/integral_set.h:381]: Hmm yeah it looks like pointer should be `DomainT*` and reference `DomainT&`

File: src/vt/vrt/collection/balance/lb_common.h (hunk 2 above)
Comment (nmm0) [reply src/vt/vrt/collection/balance/lb_common.h:82]: Ok funny story, I apparently I had a C++20 flag from a dependency. That was causing the issue, though it _shouldn't_ have. Having a deleted constructor doesn't count as a user-defined constructor in C++17 or C++20. Anyway, I'll revert that change since it isn't necessary for this PR though it does fail in C++20 mode

**Comment by nmm0** [comment]
> > > Lol at those compiler errors
> > 
> > 
> > facepalm maybe I should have tested with C++14.... well I guess I'm going to have to find another way. Maybe the least gross way will be to just do explicit capture for everything, no default capture
> 
> Or wait a bit, and we can fold in "require C++17" with the changes to adapt to it. EMPIRE will be able to accept that in the next few weeks, BVH is already there as I understand it, and I can check on our other app.

Yeah that is my bad, I was getting the C++20 flag from Kokkos. I forgot implicit this is deprecated in 20, not 17. I'll remove it from the PR

**Comment by PhilMiller** [comment]
Incidentally, you're "I'll revert that" are a good argument for committing distinct changes separately - it would be really easy to just rebase to put the `iterator` change we're keeping first, push that on this branch, and push the rest on a C++20 branch.

We'll want to figure out how to write code that's happy compiled as C++17 or C++20 starting once we require 17.

File: src/vt/termination/interval/integral_set.h [hunk 3]
```diff
@@ -378,9 +378,7 @@ struct IntegralSetBase {
   // using reference         = DomainT&;
 
   template <typename Impl>
-  struct IntervalSetIter :
-    std::iterator<std::bidirectional_iterator_tag, DomainT, DomainT>
-  {
+  struct IntervalSetIter {
```
Comment (PhilMiller) [inline src/vt/termination/interval/integral_set.h:381]: Could you also open an issue about the type aliases here being suspicious?

**Review by PhilMiller — APPROVED** [review]

**Comment by nmm0** [comment]
> Incidentally, you're "I'll revert that" are a good argument for committing distinct changes separately - it would be really easy to just rebase to put the `iterator` change we're keeping first, push that on this branch, and push the rest on a C++20 branch.
> 
> We'll want to figure out how to write code that's happy compiled as C++17 or C++20 starting once we require 17.

Split into two commits, made a third one for implicit `this` on the C++20 deprecations branch: https://github.com/DARMA-tasking/vt/tree/1976-address-cpp20-deprecations

File: src/vt/termination/interval/integral_set.h (hunk 1 above)
Comment (nmm0) [reply src/vt/termination/interval/integral_set.h]: https://github.com/DARMA-tasking/vt/issues/1977

File: src/vt/termination/interval/integral_set.h (hunk 3 above)
Comment (nmm0) [reply src/vt/termination/interval/integral_set.h:381]: https://github.com/DARMA-tasking/vt/issues/1977

**Comment by nmm0** [comment]
This PR is broken because our checking scripts don't allow '+' in the name of the branch. New PR: https://github.com/DARMA-tasking/vt/pull/1978
